File: polyglot/c/packer_detector.c
/*
 * binhunt/polyglot/c/packer_detector.c
 * 
 * Packer/Obfuscator Detector Module
 * 
 * Detects common packers and obfuscators by analyzing:
 * - ELF header signatures and sections
 * - PE headers and resource strings  
 * - File entropy (compression indicators)
 * - Known packer string patterns
 */

#define _GNU_SOURCE
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <stdint.h>
#include <math.h>
#include <sys/stat.h>
#include <fcntl.h>
#include <unistd.h>

/* Configuration constants */
#define MAX_FILE_SIZE    (1024 * 1024)      /* 1MB max for scanning */
#define ENTROPY_THRESHOLD 7.5                /* High entropy threshold */
#define STRING_SCAN_LEN  64                  /* Max string length to scan */

/* Packer signatures database */
typedef struct {
    const char *name;
    size_t len;
} PackerSignature;

static PackerSignature PACKER_SIGNATURES[] = {
    {"UPX", 3},
    {"ASPack", 6},
    {"PECompact", 9},
    {"Themida", 7},
    {"VMProtect", 10},
    {"Enigma Protector", 18},
    {"Armadillo", 10},
    {"FSG", 3},
    {"Thew", 4},
    {"MPRESS", 6},
    {"PESimp", 7},
    {"PECompact2", 11},
    {"ASPack2", 7},
    {"Themida2", 9},
    {NULL, 0}
};

/* ELF magic bytes */
#define ELF_MAGIC "\x7fELF"

/* PE signature */
#define PE_SIGNATURE "MZ"

/* File header info */
typedef struct {
    char *data;
    size_t len;
    int is_elf;
    int is_pe;
} FileHeaderInfo;

/* Detection result */
typedef struct {
    int detected;
    char packer[64];
    float entropy;
    int elf_magic;
    int pe_magic;
} DetectResult;

/* Calculate Shannon entropy of data */
static double calculate_entropy(const uint8_t *data, size_t len)
{
    if (len == 0) return 0.0;
    
    unsigned char freq[256] = {0};
    int i;
    
    /* Count byte frequencies */
    for (i = 0; i < (int)len; i++) {
        freq[data[i]]++;
    }
    
    /* Calculate entropy */
    double entropy = 0.0;
    double total = (double)len;
    
    for (i = 0; i < 256; i++) {
        if (freq[i] > 0) {
            double p = (double)freq[i] / total;
            entropy -= p * log2(p);
        }
    }
    
    return entropy;
}

/* Check ELF header for packer indicators */
static int check_elf_packer(const uint8_t *data, size_t len)
{
    if (len < 64) return 0;
    
    /* Check for common packed ELF section names */
    const char *packed_sections[] = {
        ".upx", ".upx0", ".upx1", ".upx2", ".upx3",
        ".packer", ".packed", ".compressed", ".asm_pack",
        ".themida", ".vmprotect", ".enigma",
        NULL
    };
    
    int i, j;
    for (i = 0; packed_sections[i] != NULL; i++) {
        const char *sec = packed_sections[i];
        size_t sec_len = strlen(sec);
        
        /* Search in ELF header and first few sections */
        for (j = 0; j < len - sec_len && j < 4096; j += 16) {
            if (memcmp(&data[j], sec, sec_len) == 0) {
                return 1;
            }
        }
    }
    
    /* Check for UPX magic in header */
    if (len >= 64 && memcmp(data + 52, "UPX!", 4) == 0) {
        return 1;
    }
    
    return 0;
}

/* Check PE header for packer indicators */
static int check_pe_packer(const uint8_t *data, size_t len)
{
    if (len < 64) return 0;
    
    /* Check MZ signature */
    if (memcmp(data, "MZ", 2) != 0) return 0;
    
    /* Check for PE header offset and magic */
    if (len >= 64 && data[62] == 'P' && data[63] == 'E') {
        /* Search for packer signatures in PE headers */
        const char *pe_packer_strings[] = {
            "UPX0", "UPX1", "UPX2", "UPX3",
            ".upx", ".packer", ".packed",
            "PECompact", "ASPack", "Themida",
            "VMProtect", "Enigma",
            NULL
        };
        
        int i, j;
        for (i = 0; pe_packer_strings[i] != NULL; i++) {
            const char *str = pe_packer_strings[i];
            size_t str_len = strlen(str);
            
            /* Search in PE header area */
            for (j = 64; j < len - str_len && j < 1024; j += 8) {
                if (memcmp(&data[j], str, str_len) == 0) {
                    return 1;
                }
            }
        }
    }
    
    /* Check for UPX PE header magic */
    if (len >= 64 && data[52] == 'U' && data[53] == 'P' && 
        data[54] == 'X' && data[55] == '!') {
        return 1;
    }
    
    return 0;
}

/* Scan for known packer strings throughout file */
static int scan_packer_strings(const uint8_t *data, size_t len)
{
    int i, j;
    
    /* Search for common packer name patterns */
    const char *patterns[] = {
        "UPX", "ASPack", "PECompact", "Themida", 
        "VMProtect", "Enigma Protector", "Armadillo",
        "FSG", "Thew", "MPRESS", "PESimp",
        NULL
    };
    
    for (i = 0; patterns[i] != NULL; i++) {
        const char *pat = patterns[i];
        size_t pat_len = strlen(pat);
        
        /* Case-insensitive search */
        for (j = 0; j <= len - pat_len; j++) {
            int match = 1;
            size_t k;
            
            for (k = 0; k < pat_len && match; k++) {
                if (tolower((unsigned char)data[j + k]) != 
                    tolower((unsigned char)pat[k])) {
                    match = 0;
                }
            }
            
            if (match) return 1;
        }
    }
    
    /* Check for UPX! magic in any location */
    for (j = 0; j <= len - 4; j++) {
        if (data[j] == 'U' && data[j+1] == 'P' && 
            data[j+2] == 'X' && data[j+3] == '!') {
            return 1;
        }
    }
    
    return 0;
}

/* Detect Mach-O specific packers */
static int check_macho_packer(const uint8_t *data, size_t len)
{
    if (len < 24) return 0;
    
    /* Check for Mach-O magic */
    if ((data[0] == 'f' && data[1] == 'a' && 
         data[2] == 't' && data[3] == 'h') ||
        (data[0] == 'c' && data[1] == 'o' && 
         data[2] == 'f' && data[3] == 'f')) {
        
        /* Search for packer indicators in Mach-O */
        const char *macho_patterns[] = {
            ".upx", ".packer", ".packed",
            "UPX0", "UPX1", "UPX2", "UPX3",
            NULL
        };
        
        int i, j;
        for (i = 0; macho_patterns[i] != NULL; i++) {
            const char *pat = macho_patterns[i];
            size_t pat_len = strlen(pat);
            
            for (j = 0; j <= len - pat_len; j++) {
                if (memcmp(&data[j], pat, pat_len) == 0) {
                    return 1;
                }
            }
        }
    }
    
    return 0;
}

/* Main detection function */
static int detect_packer(const uint8_t *data, size_t len, DetectResult *result)
{
    memset(result, 0, sizeof(DetectResult));
    
    /* Calculate entropy first - quick check for compression */
    result->entropy = calculate_entropy(data, len);
    
    /* Check file type and run appropriate detectors */
    if (len >= 64 && memcmp(data, ELF_MAGIC, 4) == 0) {
        result->elf_magic = 1;
        if (check_elf_packer(data, len)) {
            result->detected = 1;
            strncpy(result->packer, "UPX/ELF", sizeof(result->packer));
            return 1;
        }
    } else if (len >= 64 && memcmp(data, PE_SIGNATURE, 2) == 0) {
        result->pe_magic = 1;
        if (check_pe_packer(data, len)) {
            result->detected = 1;
            strncpy(result->packer, "UPX/PE", sizeof(result->packer));
            return 1;
        }
    } else {
        /* Generic scan for any file type */
        if (scan_packer_strings(data, len)) {
            result->detected = 1;
            strncpy(result->packer, "Generic/String", sizeof(result->packer));
            return 1;
        }
    }
    
    /* Check Mach-O as fallback */
    if (!result->elf_magic && !result->pe_magic) {
        check_macho_packer(data, len);
    }
    
    /* High entropy might indicate packing even without magic */
    if (result->entropy > ENTROPY_THRESHOLD && result->detected == 0) {
        result->detected = 1;
        strncpy(result->packer, "HighEntropy", sizeof(result->packer));
    }
    
    return result->detected;
}

/* Open and scan a file */
static int scan_file(const char *path, DetectResult *result)
{
    FILE *fp = fopen(path, "rb");
    if (!fp) {
        fprintf(stderr, "Error: Cannot open %s\n", path);
        return -1;
    }
    
    fseek(fp, 0, SEEK_END);
    long fsize = ftell(fp);
    fseek(fp, 0, SEEK_SET);
    
    if (fsize <= 0 || fsize > MAX_FILE_SIZE) {
        fclose(fp);
        return -1;
    }
    
    result->data = malloc(fsize + 1);
    if (!result->data) {
        fclose(fp);
        return -1;
    }
    
    fread(result->data, 1, fsize, fp);
    fclose(fp);
    
    /* Add null terminator for string operations */
    result->len = (size_t)fsize;
    
    return detect_packer((uint8_t*)result->data, result->len, result);
}

/* Print detection results */
static void print_result(const char *filename, DetectResult *r)
{
    printf("File: %s\n", filename ? filename : "(memory)") ;
    printf("  Detected: %s\n", r->detected ? "YES" : "NO");
    if (r->detected) {
        printf("  Packer: %s\n", r->packer);
    }
    printf("  Entropy: %.2f bits/byte\n", r->entropy);
    printf("  ELF Magic: %d, PE Magic: %d\n", r->elf_magic, r->pe_magic);
}

/* Demo/test function */
static int run_demo(void)
{
    /* Create test files in memory for demonstration */
    
    /* Test 1: Normal ELF binary (simulated header) */
    uint8_t normal_elf[64] = {
        0x7f, 'E', 'L', 'F', 2, 1, 1, 1,
        0x00, 0x00, 0x00, 0x00, 0x3e, 0x00, 0x00, 0x00,
        /* ... padding */
    };
    
    DetectResult r1;
    memset(&r1, 0, sizeof(r1));
    memcpy(r1.data, normal_elf, 64);
    r1.len = 64;
    
    int found = detect_packer((uint8_t*)r1.data, 64, &r1);
    printf("Test 1 (Normal ELF): detected=%d\n", found ? 1 : 0);
    
    /* Test 2: Simulated UPX-packed PE header */
    uint8_t upx_pe[64] = {
        'M', 'Z', 90, 3, 0, 0, 0, 0,
        0x0e, 0x1f, 0xba, 0x07, 0x00, 0x00, 0x00, 0x00,
        /* ... padding */
    };
    
    DetectResult r2;
    memset(&r2, 0, sizeof(r2));
    memcpy(r2.data, upx_pe, 64);
    r2.len = 64;
    
    found = detect_packer((uint8_t*)r2.data, 64, &r2);
    printf("Test 2 (PE Header): detected=%d\n", found ? 1 : 0);
    
    /* Test 3: High entropy data */
    uint8_t high_entropy[256];
    for (int i = 0; i < 256; i++) {
        high_entropy[i] = (uint8_t)(i * 7 + 13);
    }
    
    DetectResult r3;
    memset(&r3, 0, sizeof(r3));
    memcpy(r3.data, high_entropy, 256);
    r3.len = 256;
    
    found = detect_packer((uint8_t*)r3.data, 256, &r3);
    printf("Test 3 (High Entropy): detected=%d, entropy=%.2f\n", 
           found ? 1 : 0, r3.entropy);
    
    return 0;
}

/* Command-line interface */
static int cmd_scan(const char *path)
{
    DetectResult result;
    memset(&result, 0, sizeof(result));
    
    if (scan_file(path, &result)) {
        print_result(path, &result);
        return result.detected ? 1 : 0;
    }
    return -1;
}

/* Main entry point */
int main(int argc, char *argv[])
{
    int ret = 0;
    
    if (argc < 2) {
        /* Run demo mode */
        printf("binhunt/packer_detector: Packer Detection Module\n");
        printf("Running self-test...\n\n");
        
        ret = run_demo();
        
        printf("\nSelf-test complete.\n");
    } else {
        /* Command-line scan mode */
        for (int i = 1; i < argc && !ret; i++) {
            if (cmd_scan(argv[i])) {
                ret = 1;
            }
        }
    }
    
    return ret ? 1 : 0;
}

/* 
 * Usage: packer_detector [file1] [file2] ...
 * 
 * If no files provided, runs self-test with demo data.
 */

File: polyglot/java/packer_detector.java
package polyglot.java;

import java.io.*;
import java.nio.*;
import java.nio.channels.Channels;
import java.util.*;
import java.util.stream.Collectors;

/**
 * Packer Detector for binhunt tool.
 * Detects common packers/obfuscators by scanning magic bytes, strings, entropy patterns, and headers.
 */
public class PackerDetector {

    // Known packer signatures (magic bytes, strings)
    private static final Map<String, byte[]> PACKER_SIGNATURES = new HashMap<>();
    
    // Common packer name strings to search for
    private static final Set<String> PACKER_NAMES = new HashSet<>(Arrays.asList(
        "UPX", "UPX16", "UPX20", "Themida", "ASPack", "PECompact", 
        "VMProtect", "Enigma", "Armadillo", "PESymantec", "ThnProtect",
        "CrypTic", "Nisus", "Mars", "VirusTotal"
    ));

    // Entropy thresholds (Shannon entropy 0-8)
    private static final double HIGH_ENTROPY_THRESHOLD = 7.5;
    private static final double PACKED_SECTION_THRESHOLD = 7.2;

    public record ScanResult(
        boolean isPacked,
        List<PackerMatch> matches,
        double overallEntropy,
        String detectedPacker,
        int suspiciousSectionsCount,
        byte[] sampleData
    ) {}

    @SuppressWarnings("unchecked")
    public static void main(String[] args) throws Exception {
        // Demo: scan a real binary if provided, otherwise use embedded test data
        File target = new File(args.length > 0 ? args[0] : "test_binary.bin");
        
        if (!target.exists()) {
            System.out.println("No target file. Using embedded test pattern.");
            byte[] testData = createTestPattern();
            target = new File("test_binary.bin");
            try (FileOutputStream out = new FileOutputStream(target)) {
                out.write(testData);
            }
        }

        ScanResult result = scanBinary(target, 4096); // 4KB sample for demo
        
        System.out.println("\n=== PACKER DETECTION RESULTS ===");
        System.out.printf("File: %s\n", target.getAbsolutePath());
        System.out.printf("Overall Entropy: %.2f\n", result.overallEntropy());
        System.out.printf("Is Packed: %b\n", result.isPacked());
        System.out.printf("Suspicious Sections: %d\n", result.suspiciousSectionsCount());
        
        if (result.detectedPacker() != null) {
            System.out.println("Detected Packer: " + result.detectedPacker());
        }

        if (!result.matches().isEmpty()) {
            System.out.println("\nMatches Found:");
            for (PackerMatch m : result.matches()) {
                System.out.printf("  - %s at offset 0x%08X\n", 
                    m.name(), Long.toHexString(m.offset()));
            }
        }

        if (result.isPacked() && result.suspiciousSectionsCount() > 0) {
            System.out.println("\nRecommendation: Run baseline diff or quarantine for analysis.");
        } else {
            System.out.println("\nStatus: Appears to be a normal/unpacked binary.");
        }

        // Cleanup test file if created in memory
        if (target.getName().equals("test_binary.bin") && target.length() < 1024) {
            target.delete();
        }
    }

    /**
     * Main scanning entry point.
     */
    public static ScanResult scanBinary(File binary, int sampleSize) throws IOException {
        long fileSize = binary.length();
        byte[] data;
        
        if (sampleSize > 0 && fileSize < sampleSize) {
            // Read entire file for small files
            try (FileInputStream fis = new FileInputStream(binary)) {
                data = fis.readAllBytes();
            }
        } else {
            // Stream-based sampling for large files
            try (RandomAccessFile raf = new RandomAccessFile(binary, "r");
                 InputStream is = Channels.newChannel(raf.getChannel())
                         .mapToByteBuffer()
                         .slice(fileSize - sampleSize, sampleSize);) {
                data = is.array();
            }
        }

        // Run all detection passes
        List<PackerMatch> matches = new ArrayList<>();
        
        // Pass 1: Magic byte and signature scanning
        matches.addAll(scanMagicBytes(data));
        
        // Pass 2: String-based packer name detection
        matches.addAll(scanPackerNames(data));
        
        // Pass 3: Entropy analysis of sections (PE format)
        double entropy = calculateEntropy(data);
        List<PackerMatch> entropyMatches = scanEntropyPatterns(data, entropy);
        matches.addAll(entropyMatches);
        
        // Pass 4: Header field inspection for PE files
        if (isPEFile(data)) {
            matches.addAll(scanPEHeaders(data));
        }

        // Determine primary detected packer
        String detectedPacker = detectPrimaryPacker(matches, data.length);
        
        // Calculate suspicious section count
        int suspiciousSections = countSuspiciousSections(entropyMatches);
        
        return new ScanResult(
            matches.size() > 0 || entropy >= HIGH_ENTROPY_THRESHOLD || suspiciousSections > 2,
            matches,
            entropy,
            detectedPacker,
            suspiciousSections,
            data.length < sampleSize ? data : null // Return full data only for small files
        );
    }

    /**
     * Scan for known packer magic bytes and signatures.
     */
    private static List<PackerMatch> scanMagicBytes(byte[] data) {
        List<PackerMatch> matches = new ArrayList<>();
        
        // UPX signature: "UPX!" at offset 0x18 in PE headers (common location)
        if (data.length > 0x20) {
            byte[] upxSig = new byte[]{(byte)'U', (byte)'P', (byte)'X', '!'};
            int pos = data.indexOf(upxSig, 0x18); // Start after PE header
            if (pos >= 0x18 && pos < Math.min(data.length, 0x20 + upxSig.length)) {
                matches.add(new PackerMatch("UPX", pos));
            }
        }

        // Generic high-entropy region detection in first 64KB
        if (data.length > 65536) {
            byte[] head = Arrays.copyOfRange(data, 0, 65536);
            double headEntropy = calculateEntropy(head);
            if (headEntropy >= PACKED_SECTION_THRESHOLD) {
                matches.add(new PackerMatch("High-Entropy Header", 0));
            }
        }

        return matches;
    }

    /**
     * Scan for packer name strings embedded in the binary.
     */
    private static List<PackerMatch> scanPackerNames(byte[] data) {
        List<PackerMatch> matches = new ArrayList<>();
        
        // Search for known packer names (case-insensitive, partial match ok)
        for (String name : PACKER_NAMES) {
            int pos = -1;
            
            // Try exact match first
            byte[] bytes = name.getBytes(StandardCharsets.US_ASCII);
            pos = data.indexOf(bytes);
            
            if (pos >= 0) {
                matches.add(new PackerMatch(name, pos));
                continue;
            }

            // Try case-insensitive search in reasonable range
            for (int start = 0x1000; start < Math.min(data.length - name.length(), 0x20000); 
                 start += 4096) {
                byte[] sub = Arrays.copyOfRange(data, start, start + name.length());
                if (isSubstring(sub, bytes)) {
                    matches.add(new PackerMatch(name, start));
                    break;
                }
            }
        }

        return matches;
    }

    /**
     * Calculate Shannon entropy of byte array.
     */
    private static double calculateEntropy(byte[] data) {
        if (data.length == 0) return 0.0;
        
        int[] freq = new int[256];
        for (byte b : data) {
            freq[b & 0xFF]++;
        }

        double entropy = 0.0;
        double total = data.length;
        
        for (int count : freq) {
            if (count > 0) {
                double p = (double) count / total;
                entropy -= p * Math.log(p);
            }
        }

        return entropy;
    }

    /**
     * Scan data for entropy anomalies that suggest packing.
     */
    private static List<PackerMatch> scanEntropyPatterns(byte[] data, double overallEntropy) {
        List<PackerMatch> matches = new ArrayList<>();
        
        // Check if file is PE format and analyze sections
        if (isPEFile(data)) {
            int peOffset = 0x40; // Standard PE header offset
            
            if (data.length > peOffset + 2) {
                short machineType = getShort(data, peOffset);
                
                // Machine type 0x14C = x86, 0x8664 = x64
                boolean isX86OrX64 = (machineType == 0x14C || machineType == 0x8664);
                
                if (isX86OrX64) {
                    // Check for packed flag in PE header
                    int flagsOffset = peOffset + 60;
                    if (data.length > flagsOffset) {
                        int flags = getInt(data, flagsOffset);
                        
                        // Packed bit is bit 31 (0x80000000)
                        if ((flags & 0x80000000) != 0) {
                            matches.add(new PackerMatch("PE Header: Packed Flag", peOffset));
                        }
                    }
                }
            }
        }

        // Check for high-entropy data regions (common in packed files)
        int chunkSize = 4096;
        double chunkThreshold = PACKED_SECTION_THRESHOLD - 0.3; // Slightly lower for chunks
        
        for (int i = 0x1000; i < Math.min(data.length, 0x20000); i += chunkSize) {
            byte[] chunk = Arrays.copyOfRange(data, i, i + chunkSize);
            double chunkEntropy = calculateEntropy(chunk);
            
            if (chunkEntropy >= chunkThreshold && overallEntropy < HIGH_ENTROPY_THRESHOLD) {
                matches.add(new PackerMatch("High-Entropy Chunk", i));
            }
        }

        return matches;
    }

    /**
     * Scan PE headers for additional packer indicators.
     */
    private static List<PackerMatch> scanPEHeaders(byte[] data) {
        List<PackerMatch> matches = new ArrayList<>();
        
        if (!isPEFile(data)) {
            return matches;
        }

        int peOffset = 0x40;
        int headerSize = getShort(data, peOffset + 6); // PE header size
        
        if (data.length < peOffset + headerSize) {
            return matches;
        }

        // Check for common packer-modified fields
        int flags = getInt(data, peOffset + 60);
        
        // Bit 31: Packed flag
        if ((flags & 0x80000000) != 0) {
            matches.add(new PackerMatch("PE Header", peOffset, "Packed Flag Set"));
        }

        // Check section headers for obfuscated names or high entropy
        int numSections = getShort(data, peOffset + 2);
        int sectionsStart = peOffset + 64;
        
        if (data.length >= sectionsStart + numSections * 40) {
            for (int i = 0; i < numSections && i < 100; i++) {
                int sectionNameOffset = sectionsStart + i * 40;
                
                // Extract first few bytes of section name
                byte[] nameBytes = Arrays.copyOfRange(data, 
                    sectionNameOffset, Math.min(sectionNameOffset + 8, data.length));
                
                String name = new String(nameBytes, StandardCharsets.US_ASCII).trim();
                
                // Check for obfuscated/short names (common in packed files)
                if (name.isEmpty() || name.length() < 3 || 
                    name.length() > 15 && !isReadableName(name)) {
                    matches.add(new PackerMatch("PE Header", sectionNameOffset, 
                        "Suspicious Section Name: \"" + name + "\""));
                }

                // Check section characteristics
                int flags = getInt(data, sectionNameOffset + 2);
                
                // Read-only and compressed sections are suspicious
                if ((flags & 0x80000000) != 0 || (flags & 0x40000000) != 0) {
                    matches.add(new PackerMatch("PE Header", sectionNameOffset, 
                        "Compressed/RO Section"));
                }

                // Check for high entropy in code sections
                int virtualSize = getInt(data, sectionNameOffset + 16);
                if (virtualSize > 0 && virtualSize < data.length) {
                    byte[] sectionData = Arrays.copyOfRange(
                        data, 
                        Math.max(peOffset + headerSize, sectionNameOffset),
                        Math.min(sectionNameOffset + virtualSize, data.length));
                    
                    double secEntropy = calculateEntropy(sectionData);
                    if (secEntropy >= PACKED_SECTION_THRESHOLD) {
                        matches.add(new PackerMatch("PE Header", sectionNameOffset, 
                            "High-Entropy Section"));
                    }
                }
            }
        }

        return matches;
    }

    /**
     * Determine the primary detected packer from all matches.
     */
    private static String detectPrimaryPacker(List<PackerMatch> matches, int fileSize) {
        // Priority order for detection confidence
        if (matches.stream().anyMatch(m -> m.name().equals("UPX"))) {
            return "UPX";
        }

        if (matches.stream().anyMatch(m -> m.name().contains("PE Header: Packed Flag"))) {
            return "Unknown PE Packer";
        }

        // Check for specific packer strings with high confidence
        String[] highConfidence = {"VMProtect", "Themida", "Armadillo"};
        for (String p : highConfidence) {
            if (matches.stream().anyMatch(m -> m.name().equals(p))) {
                return p;
            }
        }

        // Fallback: most frequent match name
        Map<String, Integer> freq = matches.stream()
            .collect(Collectors.groupingBy(PackerMatch::name, Collectors.counting()));
        
        if (!freq.isEmpty()) {
            String topPacker = freq.entrySet().stream()
                .max(Comparator.comparingInt(Map.Entry::getValue))
                .map(Map.Entry::getKey)
                .orElse(null);
            
            // Only return if it's a known packer and appears multiple times
            if (topPacker != null && topPacker.contains("PE") && freq.get(topPacker) >= 2) {
                return "Unknown PE Packer";
            }
        }

        return matches.isEmpty() ? null : 
               matches.stream().map(PackerMatch::name).distinct().limit(3).collect(Collectors.joining(", "));
    }

    /**
     * Count sections with suspicious characteristics.
     */
    private static int countSuspiciousSections(List<PackerMatch> entropyMatches) {
        Set<Integer> suspiciousOffsets = new HashSet<>();
        
        for (PackerMatch m : entropyMatches) {
            if (m.name().contains("PE Header") || 
                m.name().contains("High-Entropy")) {
                int offset = m.offset() & 0xFFFF; // PE sections are at offsets like 0x1000, 0x2000
                
                // Map section header offsets to their virtual addresses
                // Section headers start after the main PE header
                int peHeaderSize = getShort(new byte[]{(byte)0x4D, (byte)'P', (byte)'E'}, 6);
                
                if (offset > 0x1000 && offset < 0x20000) {
                    suspiciousOffsets.add(offset &